core/cap-0051.md

```diff
@@ -260,18 +260,22 @@ func $serialize_to_binary(param $val i64) (result i64)
 /// Traps if the deserialization fails for any reason.
 func $deserialize_from_binary(param $obj_bin i64, param $type u64) (result i64)
 
-/// Given a host binary object $obj_bin, copy a segment of its memory specified at $offset with length $len into guest linear memory at $pos. 
+/// Given a host binary object $obj_bin, copies a segment of its memory specified at $offset with length $len into the guest linear memory at $pos. 
 /// Traps if either the binary object or the guest linear memory doesn't have enough bytes.
-func $binary_copy_to_guest_mem(param $obj_bin i64) (param $offset u64) (param $pos u64) (param $len u64)
+func $binary_memcpy_to_guest(param $obj_bin i64) (param $offset u64) (param $pos u64) (param $len u64)
 
-/// Copy a segment of the guest's linear memory specified at position $pos with length $len, into a host binary object $obj_bin at $offset. 
+/// Copies a segment of the guest linear memory specified at position $pos with length $len, into a host binary object $obj_bin at $offset. The host binary binary may grow in size to accomadate the new bytes. 
 /// Returns handle to the new binary array.
-/// Traps if either the binary object or the guest linear memory doesn't have enough bytes.
-func $binary_copy_from_guest_mem(param $obj_bin i64) (param $offset u64) (param $pos u64) (param $len u64) (result i64)
+/// Traps if the guest linear memory doesn't have enough bytes.
+func $binary_memcpy_from_guest(param $obj_bin i64) (param $offset u64) (param $pos u64) (param $len u64) (result i64)
 
 /// Construct an empty binary array, returns its handle
 func $binary_new() (result i64)
 
+/// Constructs a new binary array and initializes it with bytes copied from a guest linear memory slice specified at position $pos with length $len. 
+/// Returns handle to the new binary array.
+func $binary_new_from_memcpy(param $pos u64) (param $len u64) (result i64)
```

Comment:
💡 I don't feel strongly, but might I suggest the following names instead.

The term memcpy is removed because it feels out of place. It is a highly specific term from specific operating systems and the C/C++ stdlibs.

The term guest is helpful to understand what the source / destination is, but I think we can be more specific and say linear memory instead. If you read articles and documentation about WASM about memory interactions the guests memory is always referred to as the linear memory, and so it makes the names of these functions aligned with the types of things we'd be searching for that is relevant.

As a bonus these names help align the functions a little more than they are today.

```wat
func $binary_copy_to_linear_memory (param $obj_bin i64) (param $offset u64) (param $pos u64) (param $len u64)
func $binary_new () (result i64)
func $binary_new_from_linear_memory (param $pos u64) (param $len u64) (result i64)
```